streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -339,28 +378,36 @@ private void closeTopology() {
 
     /**
      * <pre>
-     *  - {@link #suspend()}
-     *    - close topology
-     *    - {@link #commit()}
-     *      - flush state and producer
-     *      - write checkpoint
-     *      - commit offsets
-     *  - close state
+     * - {@link #suspend(boolean) suspend(clean)}
+     *   - close topology
+     *   - if (clean) {@link #commit(boolean) commit(noNewTransaction)}
+     *     - flush state and producer
+     *     - if (!eos) write checkpoint
+     *     - commit offsets
+     * - close state
+     *   - if (clean) write checkpoint
      * </pre>
      */
     @Override
-    public void close() {
+    public void close(final boolean clean) {
         log.debug("{} Closing", logPrefix);
         try {
-            suspend();
-            closeStateManager(true);
+            suspend(clean);
+            closeStateManager(clean);
             partitionGroup.close();
             metrics.removeAllSensors();
         } catch (final RuntimeException e) {
             closeStateManager(false);
             throw e;
         } finally {
             if (exactlyOnceEnabled) {
+                if (!clean) {
+                    try {
+                        producer.abortTransaction();
```

Comment:
In case of unclean shutdown, we need to roll back and should not commit the transaction.